README
Some common notes/FAQs:
======================

* when vserver startup/shutdown fails, or when you get

  | Error: /proc must be mounted

  errors, make sure, that 'vprocunhide' was executed. When installing
  'util-vserver' with packagemanagement, an appropriate initscript
  should be installed

* the name of old-style vservers is shown on 2.4 kernels only; the
  needed functionality is not implemented for 2.6 kernels.



Some distribution specific notes:
=================================

Red Hat 7.3, Red Hat 9, Fedora Core 1&2
---------------------------------------
* tested and running successfully as host and guest systems

* it is *strongly* suggested to use the rpm packages which can be
  created from the tarball with

  | $ rpmbuild -tb util-vserver-<version>.tar.bz2

  For distributions below Fedora Core 2, additional

  | --without dietlibc --without xalan

  flags are required for the 'rpmbuild' command. Builds on Red Hat 7.3
  will require a

  | --nodeps

  also, since 'vconfig' is not available there. Since it is required
  for path-detection only and paths from RH systems will be assumed by
  default, this should not be a big problem.

* guest systems can be created with the 'apt-rpm' or 'yum' build-methods.
  The first one requires the 'apt' package e.g. from http://fedora.us and
  the configuration of a near mirror in

  | /etc/vservers/.distributions/<id>/apt/sources.list

  (To avoid slashdotting by the masses of util-vserver-users, there
  does not exist a standard mirror).

  The 'yum' method uses the repository configuration shipped by the
  fedora-release package.

* RH/FC uses the 'sysv' initstyle which is assumed by default

* when having existing vservers with RH 9 or Fedora Core 1, the startup
  of the vserver will probably fail. You will have to add

  | true

  to etc/rc.d/rc (within the vserver root directory)

* when having RH/FC guestsystems, it is *strongly* recommended to use
  a dietlibc linked version of 'rpm-fake-resolver'. Else, package
  installation with 'vrpm', 'vapt-get' or 'vyum' can fail since users
  can not be resolved.



Debian Woody & Sarge
--------------------
* tested and running successfully as guest systems on FC1/FC2 hosts

* guest systems can be created with the 'debootstrap' method. When
  not already existing, the needed package will be downloaded
  automatically.  Since it is updated very often, it can happen
  that a '404 Not found' error occurs; in this case look either
  for a newer util-vserver package, or configure the new URI e.g. with

  | echo 'http://ftp.debian.org/debian/pool/main/d/debootstrap/debootstrap_<version>_i386.deb' \
  |    >/etc/vservers/.defaults/apps/debootstrap/uri

  You can download a local copy of this tarball also, and register it
  with

  | echo '/<path-to-the-tarball>' \
  |    >/etc/vservers/.defaults/apps/debootstrap/uri

* it is known, that warning messages will be created at startup and
  shutdown of guest servers. This is non fatal and can be ignored

* Debian guest systems are running fine with the 'sysv' initstyle;
  success with 'plain' was reported also

* no packages for Debian hosts are known at time of writing (May 2004)



Gentoo
------
* Gentoo guest systems are very complicated and are requiring lots of
  modifications in the initscripts. Currently, no step-by-step guide
  can be provided

* 'sysv' initstyle is probably not working for Gentoo guests (e.g. you
  will see messages about missing 'utmp' files); 'gentoo' should be
  used instead of:

  | echo 'gentoo' >/etc/vservers/<id>/apps/init/style

* there does not exist a build-method for Gentoo guests; instead of,
  create a skeleton with

  | # vserver <id> build -m skeleton --initstyle gentoo <other-opts>*

  and fill the vserver directory at /etc/vservers/<id>/vdir/ manually.



Notes for distributors:
=======================

To generate FHS compliant paths, call configure with

| ./configure --prefix=/usr --mandir=/usr/share/man \
|             --sysconfdir=/etc --localstatedir=/var \
|             --with-vrootdir=<an FHS compliant path for /vservers>

Except the '--with-vrootdir' option, rpm's '%configure' option will
expand to this.


There exists a 'make install-distribution' target which installs
files outside of the configured 'prefix'. In particular, these files are:

* the /sbin/vshelper symlink
* the /vservers and related directories (or whatever you configured
  with '--with-vrootdir')

Without this rule, 'make distcheck' would fail.


It might be needed also, to call 'setattr --barrier /vservers' in an
after-installation script.



Which version shall I use?
==========================

As you probably know, two branches of 'util-vserver' are existing: the
'stable' one, and the 'alpha' one. This terms are to be understood as
a level of the featureset stability but not of the software stability.

E.g. 'stable' is not really stable: it has huge security problems and
missing functionality. But you can expect that the current configuration
will work in future versions also. This version is untested on author's
side and it will be hard to bring patches/fixes in, since it must be
proofed that they will not break anything.

In the opposite, the 'alpha' branch does not have known security issues
and works well (at least on author's system ;)).  But it may happen
that some behavior or configuration options change.

With 'alpha' you should be still able to use vservers created with the
'stable' branch, but you may encounter some oddities -- especially on
kernel 2.6 systems (e.g. 'vserver-stat' will not show the names of old
vservers).


So let me summarize:

* when you have productive vservers running for some years already, stay
  at the 'stable' branch. A change to 'alpha' will need a completely
  rewritten configuration which must be perhaps changed again.

* when you are new at vservers, use the 'alpha' branch. You will have
  to learn the principles of vserver configuration for both branches
  but 'alpha' makes some things easier.

* when you have existing vservers and want all the new kernel 2.6
  functionality, use the 'alpha' branch.


A last note: the 'alpha' branch works both with the stable 2.4 and the
development 2.6 kernel patch.



## $Id: README,v 1.7 2005/01/27 19:07:05 ensc Exp $


ChangeLog
2009-04-10 20:24  Daniel Hokka Zakrisson <daniel@hozac.com>

	* src/secure-mount.c: Handle shared/private/unbindable/slave mounts
	  internally.

2009-04-03 16:41  Daniel Hokka Zakrisson <daniel@hozac.com>

	* src/secure-mount.c: The remount still needs to have MS_BIND set.

2009-03-18 19:26  Daniel Hokka Zakrisson <daniel@hozac.com>

	* scripts/vserver.functions: -o shared doesn't work with --rbind.

2009-02-14 10:07  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* m4/ensc_dietlibc.m4: whitespace cleanup

2009-02-14 10:06  Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>

	* m4/ensc_dietlibc.m4: fixed detection of dietlibc version strings
	  like 0.31-1.2

2008-12-01 20:48  Daniel Hokka Zakrisson <daniel@hozac.com>

	* scripts/vserver.suexec: Attach to the cgroup.

2008-11-11 20:59  Daniel Hokka Zakrisson <daniel@hozac.com>

	* distrib/defaults/devs: Add /dev/shm.

2008-11-11 20:58  Daniel Hokka Zakrisson <daniel@hozac.com>

	* lib/getinsecurebcaps.c: Remove the #ifdef since the logic got
	  inverted.

2008-11-06 17:13  Daniel Hokka Zakrisson <daniel@hozac.com>

	* ., COPYING: Add COPYING to svn.

2008-11-06 02:27  Daniel Hokka Zakrisson <daniel@hozac.com>

	* configure.ac: Add a way to disable library versioning. Better
	  than nothing until proper checks are written.

2008-11-06 02:18  Daniel Hokka Zakrisson <daniel@hozac.com>

	* python/libvserver.py: Add missing piece of code.
	  Fix __add__ and __sub__ in class struct.

2008-11-06 02:00  Daniel Hokka Zakrisson <daniel@hozac.com>

	* python/_libvserver.c, python/libvserver.py: Add Pythonizing
	  wrappers for the Python-bindings.

2008-11-03 19:52  Daniel Hokka Zakrisson <daniel@hozac.com>

	* scripts/chcontext: Enter the spaces when migrating to an existing
	  static context.

2008-10-31 15:46  Daniel Hokka Zakrisson <daniel@hozac.com>

	* lib/Makefile-files: Sigh.

2008-10-31 15:41  Daniel Hokka Zakrisson <daniel@hozac.com>

	* lib/issupported.c, lib/issupportedstring.c, lib/vserver.h,
	  scripts/vserver.functions, scripts/vserver.start,
	  scripts/vserver.stop, scripts/vserver.suexec, src/secure-mount.c,
	  src/vcontext.c: Use pivot_root on kernels supporting two
	  namespaces per context.

2008-10-31 15:40  Daniel Hokka Zakrisson <daniel@hozac.com>

	* kernel/space_cmd.h, kernel/switch.h,
	  lib/syscall_getspacemask-v21.hc, lib/syscall_getspacemask-v23.hc,
	  lib/syscall_getspacemask.c: Fix vc_get_space_default and add
	  vc_get_space_mask v1.

2008-10-31 04:45  Daniel Hokka Zakrisson <daniel@hozac.com>

	* ensc_wrappers/wrappers-vserver.hc, kernel/space_cmd.h,
	  lib/Makefile-files, lib/syscall_enternamespace-v13.hc,
	  lib/syscall_enternamespace-v21.hc,
	  lib/syscall_enternamespace-v23.hc, lib/syscall_enternamespace.c,
	  lib/syscall_getspacedefault-v23.hc,
	  lib/syscall_getspacemask-v21.hc, lib/syscall_setnamespace-v13.hc,
	  lib/syscall_setnamespace-v21.hc, lib/syscall_setnamespace-v23.hc,
	  lib/syscall_setnamespace.c, lib/vserver.h, python/_libvserver.c,
	  src/vcontext.c, src/vspace.c: Add VCMD_{enter,set}_space_v2
	  support, in preparation of pivot_root.

2008-10-31 04:21  Daniel Hokka Zakrisson <daniel@hozac.com>

	* scripts/Makefile-files, scripts/util-vserver-vars.pathsubst,
	  scripts/vnamespace, scripts/vserver-build.debootstrap,
	  scripts/vsomething, src/Makefile-files, src/vnamespace.c: Remove
	  vnamespace in favor of vspace and a small wrapper around it.

2008-10-31 04:07  Daniel Hokka Zakrisson <daniel@hozac.com>

	* lib/vserver-internal.h: Add hidden attribute to be used for
	  versioned symbols.

2008-10-31 04:05  Daniel Hokka Zakrisson <daniel@hozac.com>

	* lib/syscall_ctxcreate-v21.hc, lib/syscall_ctxmigrate-v21.hc,
	  scripts/chcontext, scripts/vserver.functions,
	  scripts/vserver.suexec: Move space creation/entering out of
	  vc_ctx_{create,migrate}.

2008-10-31 04:00  Daniel Hokka Zakrisson <daniel@hozac.com>

	* src/vspace.c: CLONE_NEWNS and CLONE_FS aren't default spaces.

2008-10-31 04:00  Daniel Hokka Zakrisson <daniel@hozac.com>

	* lib/syscall_getspacemask-v21.hc: Filter out CLONE_NEWNS and
	  CLONE_FS, as vc_get_space_default does.

2008-10-30 22:45  Daniel Hokka Zakrisson <daniel@hozac.com>

	* python/_libvserver.c: Need to be able to retrieve per-CPU values.

2008-10-30 21:23  Daniel Hokka Zakrisson <daniel@hozac.com>

	* configure.ac, lib/Makefile-files, lib/internal.h,
	  lib/libvserver.ver: Library versioning stubs.

2008-10-30 02:33  Daniel Hokka Zakrisson <daniel@hozac.com>

	* configure.ac, python/Makefile-files, python/_libvserver.c,
	  python/ctags-constants.awk: 32-bit and Python 2.4 support.

2008-10-30 02:25  Daniel Hokka Zakrisson <daniel@hozac.com>

	* lib/getinsecurebcaps.c: These got lost somehow.

2008-10-30 01:59  Daniel Hokka Zakrisson <daniel@hozac.com>

	* util-vserver.spec.in: BR: ctags

2008-10-30 01:21  Daniel Hokka Zakrisson <daniel@hozac.com>

	* configure.ac, python/Makefile-files, python/_libvserver.c,
	  python/ctags-constants.awk: Some minor cleanups in the Python
	  build.

2008-10-29 18:37  Daniel Hokka Zakrisson <daniel@hozac.com>

	* configure.ac: Just in case anything else wants to use Python as
	  part of the build process.

2008-10-29 18:36  Daniel Hokka Zakrisson <daniel@hozac.com>

	* python/_libvserver.c: 65 is the length of the command structure's
	  field.

2008-10-28 00:56  Daniel Hokka Zakrisson <daniel@hozac.com>

	* util-vserver.spec.in: Add Python stuff.

2008-10-28 00:51  Daniel Hokka Zakrisson <daniel@hozac.com>

	* contrib/manifest.dat.pathsubst: vserver.start.bin

2008-10-28 00:24  Daniel Hokka Zakrisson <daniel@hozac.com>

	* Makefile.am, configure.ac, python, python/Makefile-files,
	  python/_libvserver.c: Add Python bindings to libvserver.

2008-10-28 00:23  Daniel Hokka Zakrisson <daniel@hozac.com>

	* doc/configuration.xml, scripts/vserver-setup.functions,
	  scripts/vserver.functions: Rename flags to cflags.

2008-10-27 23:52  Daniel Hokka Zakrisson <daniel@hozac.com>


src/vcopy.c
// $Id: vcopy.c 2403 2006-11-24 23:06:08Z dhozac $    --*- c -*--

// Copyright (C) 2004 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
//  
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; version 2 of the License.
//  
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//  
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.


#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#include "util.h"
#include "vserver.h"

#include "lib_internal/matchlist.h"
#include "lib_internal/unify.h"

#include <unistd.h>
#include <getopt.h>
#include <fcntl.h>
#include <dirent.h>
#include <errno.h>
#include <assert.h>
#include <utime.h>
#include <libgen.h>
#include <sys/param.h>

#define ENSC_WRAPPERS_UNISTD	1
#define ENSC_WRAPPERS_FCNTL	1
#define ENSC_WRAPPERS_DIRENT	1
#include <wrappers.h>

#define CMD_HELP		0x8000
#define CMD_VERSION		0x8001
#define CMD_MANUALLY		0x8002
#define CMD_STRICT		0x8003

struct WalkdownInfo
{
    PathInfo				state;
    struct MatchList			dst_list;
    struct MatchList			src_list;
};

struct Arguments {
    enum {mdMANUALLY, mdVSERVER}	mode;
    bool				do_dry_run;
    unsigned int			verbosity;
    bool				local_fs;
    bool				is_strict;
};

static struct WalkdownInfo		global_info;
static struct Arguments const *		global_args;

int wrapper_exit_code = 1;

struct option const
CMDLINE_OPTIONS[] = {
  { "help",     no_argument,  0, CMD_HELP },
  { "version",  no_argument,  0, CMD_VERSION },
  { "manually", no_argument,  0, CMD_MANUALLY },
  { "strict",   no_argument,  0, CMD_STRICT },
  { 0,0,0,0 }
};

typedef enum { opUNIFY, opCOPY, opDIR, opSKIP }	Operation;

static void
showHelp(int fd, char const *cmd, int res)
{
  VSERVER_DECLARE_CMD(cmd);
  
  WRITE_MSG(fd, "Usage:\n  ");
  WRITE_STR(fd, cmd);
  WRITE_MSG(fd,
	    " [-nv] [--strict] <dst-vserver> <src-vserver>\n    or\n  ");
  WRITE_STR(fd, cmd);
  WRITE_MSG(fd,
	    " --manually [-nvx] [--] <dst-path> <dst-excludelist> <src-path> <src-excludelist>\n\n"
 	    "  --manually      ...  unify generic paths; excludelists must be generated\n"
	    "                       manually\n"
	    "  --strict        ...  require an existing vserver configuration for dst-vserver;\n"
	    "                       by default, a base skeleton will be created but manual\n"
	    "                       configuration wil be still needed to make the new vserver work\n"
	    "  -n              ...  do not modify anything; just show what there will be\n"
	    "                       done (in combination with '-v')\n"
	    "  -v              ...  verbose mode\n"
	    "  -x              ...  do not cross filesystems; this is valid in manual\n"
	    "                       mode only and will be ignored for vserver unification\n\n"
	    "Please report bugs to " PACKAGE_BUGREPORT "\n");
  exit(res);
}

static void
showVersion()
{
  WRITE_MSG(1,
	    "vcopy " VERSION " -- copies directories and vserver files\n"
	    "This program is part of " PACKAGE_STRING "\n\n"
	    "Copyright (C) 2003,2004 Enrico Scholz\n"
	    VERSION_COPYRIGHT_DISCLAIMER);
  exit(0);
}

int Global_getVerbosity() {
  return global_args->verbosity;
}

bool Global_doRenew() {
  return true;
}

#include "vserver-visitdir.hc"

static Operation
checkDirEntry(PathInfo const *path, struct stat const *st)
{
  struct WalkdownInfo const * const	info = &global_info;
  MatchType				res;

  // when marked as 'skip' in the first excludelist already, we do not need to
  // visit the second one since it could not change that.
  res=MatchList_compare(&info->dst_list, path->d);
  if (res!=stSKIP) {
    MatchType		tmp = MatchList_compare(&info->src_list, path->d);

    // stINCLUDE gets overridden by stEXCLUDE+stSKIP, and stEXCLUDE by stSKIP.
    // Using the MAX() macro is a hack but it works
    res=MAX(res,tmp);
  }

  // non-skipped directories are marked as opDIR
  if (res!=stSKIP && S_ISDIR(st->st_mode))
    return opDIR;

  // non-skipped symlinks will be copied always
  if (res!=stSKIP && S_ISLNK(st->st_mode))
    return opCOPY;

  // skipped files or non regular files (character/block devices) will be skipped
  // always
  if (res==stSKIP || !S_ISREG(st->st_mode))
    return opSKIP;

  switch (res) {
    case stINCLUDE	:  return opUNIFY;
    case stEXCLUDE	:  return opCOPY;
    case stSKIP		:  assert(false); // already handled above
    default		:  assert(false); abort();
  }
}

static bool
doit(Operation op,
     PathInfo const *dst_path,
     PathInfo const *src_path, struct stat const *exp_stat,
     PathInfo const *show_path)
{
#if 0
  struct stat		st;

  if (lstat(dst_path->d, &st)!=-1) {
    if (global_args->do_keep &&
	(!S_ISDIR(exp_stat->st_mode) || S_ISDIR(st.st_mode))) {
      // when keep-mode is enable and, do nothing
      if (global_args->do_dry_run || global_args->verbosity>1) {
	WRITE_MSG(1, "keeping  '");
	write(1, show_path->d, show_path->l);
	WRITE_MSG(1, "'\n");
      }
      return true;
    }
    
  }
#endif
  
  if (global_args->do_dry_run || global_args->verbosity>1) {
    if      (op==opUNIFY)   WRITE_MSG(1, "linking  '");
    else if (op==opCOPY)    WRITE_MSG(1, "copying  '");
    else if (op==opDIR)     WRITE_MSG(1, "creating '");
    else if (op==opSKIP)    WRITE_MSG(1, "skipping '");
    else { assert(false); abort(); }

    Vwrite(1, show_path->d, show_path->l);
    WRITE_MSG(1, "'\n");
  }

  return (global_args->do_dry_run ||
	  ( op==opSKIP) ||
	  ( op==opUNIFY && Unify_unify(src_path->d, exp_stat, dst_path->d, false)) ||
	  ((op==opCOPY  ||
	    op==opDIR)  && Unify_copy (src_path->d, exp_stat, dst_path->d)));
}
     
static uint64_t
visitDirEntry(struct dirent const *ent)
{
  char const *			dirname  = ent->d_name;
  if (isDotfile(dirname)) return 0;

  uint64_t			res      = 1;
  PathInfo			src_path = global_info.state;
  PathInfo			src_d_path = {
    .d = dirname,
    .l = strlen(dirname)
  };
  char				path_buf[ENSC_PI_APPSZ(src_path, src_d_path)];
  struct stat			f_stat = { .st_dev = 0 };

  PathInfo_append(&src_path, &src_d_path, path_buf);

  
  if (lstat(dirname, &f_stat)==-1)
    perror("lstat()");
  else {
    Operation		op       = checkDirEntry(&src_path, &f_stat);
    PathInfo		dst_path = global_info.dst_list.root;
    char		dst_path_buf[ENSC_PI_APPSZ(dst_path, src_path)];

    PathInfo_append(&dst_path, &src_path, dst_path_buf);
    if (!doit(op, &dst_path, &src_d_path, &f_stat, &src_path))
      perror(src_path.d);
    else if (op==opDIR) {
      res = visitDir(dirname, &f_stat);
      if (!global_args->do_dry_run &&
	  !Unify_setTime(dst_path.d, &f_stat))
	perror("utime()");
    }
    else
      res = 0;
  }

  return res;
}

#include "vcopy-init.hc"

int main(int argc, char *argv[])
{
  struct Arguments	args = {
    .mode		=  mdVSERVER,
    .do_dry_run		=  false,
    .verbosity		=  0,
    .local_fs		=  false,
  };
  uint64_t		res;

  global_args = &args;
  while (1) {
    int		c = getopt_long(argc, argv, "nvcx",
				CMDLINE_OPTIONS, 0);
    if (c==-1) break;

    switch (c) {
      case CMD_HELP		:  showHelp(1, argv[0], 0);
      case CMD_VERSION		:  showVersion();
      case CMD_MANUALLY		:  args.mode       = mdMANUALLY; break;
      case CMD_STRICT		:  args.is_strict  = true; break;
      case 'n'			:  args.do_dry_run = true; break;
      case 'x'			:  args.local_fs   = true; break;
      case 'v'			:  ++args.verbosity;       break;
      default		:
	WRITE_MSG(2, "Try '");
	WRITE_STR(2, argv[0]);
	WRITE_MSG(2, " --help' for more information.\n");
	return EXIT_FAILURE;
	break;
    }
  }

  if (argc==optind) {
    WRITE_MSG(2, "No directory/vserver given\n");
    return EXIT_FAILURE;
  }

  switch (args.mode) {
    case mdMANUALLY	:  initModeManually(argc-optind, argv+optind); break;
    case mdVSERVER	:  initModeVserver (argc-optind, argv+optind); break;
    default		:  assert(false); return EXIT_FAILURE;
  }

  if (global_args->verbosity>3)
    WRITE_MSG(1, "Starting to traverse directories...\n");

  Echdir(global_info.src_list.root.d);
  res = visitDir("/", 0);
  
#ifndef NDEBUG
  {
    MatchList_destroy(&global_info.dst_list);
    MatchList_destroy(&global_info.src_list);
  }
#endif

  return res>0 ? 1 : 0;
}
